pyadjoint/reduced_functional.py

```diff
@@ -41,42 +83,55 @@ def __init__(self, functional, controls, scale=1.0, tape=None,
         self.hessian_cb_pre = hessian_cb_pre
         self.hessian_cb_post = hessian_cb_post
 
+        if self.derivative_components:
+            # pre callback
+            self.derivative_cb_pre = get_extract_derivative_components(
+                derivative_components)
+            # post callback
+            self.derivative_cb_post = get_pack_derivative_components(
+                controls, derivative_components)
+
     def derivative(self, adj_input=1.0, options={}):
         """Returns the derivative of the functional w.r.t. the control.
-
         Using the adjoint method, the derivative of the functional with
-        respect to the control, around the last supplied value of the control,
-        is computed and returned.
+        respect to the control, around the last supplied value of the
+        control, is computed and returned.
 
         Args:
-            options (dict): A dictionary of options. To find a list of available options
-                have a look at the specific control type.
+            options (dict): A dictionary of options. To find a list of
+                available options have a look at the specific control type.
 
         Returns:
             OverloadedType: The derivative with respect to the control.
                 Should be an instance of the same type as the control.
 
         """
+
         # Call callback
         values = [c.tape_value() for c in self.controls]
-        self.derivative_cb_pre(self.controls.delist(values))
+        controls = self.derivative_cb_pre(self.controls)
+        if not isinstance(controls, Enlist):
+            raise DeprecationWarning("""Note that the callback interface
+            for derivative_cb_pre has changed. It should now return a
+            list of controls (usually the same list as input.""")
```

Comment:
Error message doesn't match if statement. Is it `list` or `Enlist`?

Reply:
Fixed in latest commit - I just check that the callback returns something now, because we just need to alert people to the change in interface.